### PR TITLE
chore: Update from functions to functions2

### PR DESCRIPTION
**Pull request by a5chin**
# Summary
The version of cloud functions written in terraform was 1st gen, so I updated it to 2nd gen.
`google_cloudfunctions_function`->`google_cloudfunctions2_function`

# Why you need this PR
As shown below, the 2nd generation is recommended over the 1st generation in Google Cloud.

> We recommend that you choose Cloud Functions (2nd gen) for new functions wherever possible. However, we plan to continue supporting Cloud Functions (1st gen).

https://cloud.google.com/functions/docs/concepts/version-comparison

# What I did
The following 2 files have been updated.
- `autoscalse/terraform/modules/autoscaler-functions/main.tf`
- `autoscalse/terraform/modules/forwarder/main.tf`

# FYI
Already tested with our products.

### CODE REVIEW

**Comment by alexlo03** [comment]
Can I suggest this go into V3?

**Comment by alexlo03** [comment]
@henrybell if you're looking to cut v3 please consider this one

File: terraform/modules/autoscaler-functions/main.tf [hunk 1]
```diff
@@ -96,46 +105,90 @@ resource "google_storage_bucket_object" "gcs_functions_source" {
...
 
   lifecycle {
-    ignore_changes = [max_instances]
+    ignore_changes = [
+      service_config[0].max_instance_count
+    ]
   }
 }
+
+resource "google_cloud_run_service_iam_member" "cloud_run_poller_invoker" {
```
[the remark this answers is not in the document]
Comment (nielm) [reply terraform/modules/autoscaler-functions/main.tf:178]: fixed by d20dfa3

File: terraform/modules/forwarder/main.tf [hunk 2]
```diff
@@ -90,27 +98,48 @@ resource "google_storage_bucket_object" "gcs_functions_forwarder_source" {
...
+    service_account_email = google_service_account.forwarder_sa.email
+    environment_variables = {
+      POLLER_TOPIC = var.target_pubsub_topic
+    }
+  }
+
+  event_trigger {
+    event_type   = "google.cloud.pubsub.topic.v1.messagePublished"
+    pubsub_topic = google_pubsub_topic.forwarder_topic.id
+    retry_policy = "RETRY_POLICY_RETRY"
```
[the remark this answers is not in the document]
Comment (nielm) [reply terraform/modules/forwarder/main.tf]: fixed by d20dfa3

File: terraform/modules/forwarder/main.tf [hunk 3]
```diff
@@ -90,27 +98,48 @@ resource "google_storage_bucket_object" "gcs_functions_forwarder_source" {
...
-  service_account_email = google_service_account.forwarder_sa.email
-  build_service_account = google_service_account.build_sa.id
+
   depends_on = [
     time_sleep.wait_for_iam,
     google_pubsub_topic_iam_member.forwader_pubsub_sub_binding
   ]
 }
+
+resource "google_cloud_run_service_iam_member" "cloud_run_forwarder_invoker" {
```
[the remark this answers is not in the document]
Comment (nielm) [reply terraform/modules/forwarder/main.tf:136]: fixed by d20dfa3

File: terraform/modules/autoscaler-functions/main.tf [hunk 4]
```diff
@@ -96,46 +105,90 @@ resource "google_storage_bucket_object" "gcs_functions_source" {
...
+    ingress_settings      = "ALLOW_INTERNAL_AND_GCLB"
+    service_account_email = var.poller_sa_email
+  }
+
   event_trigger {
-    event_type = "google.pubsub.topic.publish"
-    resource   = google_pubsub_topic.poller_topic.id
+    event_type   = "google.cloud.pubsub.topic.v1.messagePublished"
+    pubsub_topic = google_pubsub_topic.poller_topic.id
+    retry_policy = "RETRY_POLICY_RETRY"
```
[the remark this answers is not in the document]
Comment (nielm) [reply terraform/modules/autoscaler-functions/main.tf]: fixed by d20dfa3

**Comment by nielm** [comment]
The most recent commit 87a221c is a squash of the existing commits and an update of the commit message so that the breaking change information appears in the changelog

**Review by nielm — APPROVED** [review]
LGTM

**Comment by henrybell** [comment]
Thanks @a5chin!

**Pull request merged**